ComponentKit/Utilities/CKComponentAction.h

```diff
@@ -11,44 +11,102 @@
 #import <UIKit/UIKit.h>
 
 #import <ComponentKit/CKComponentViewAttribute.h>
+#import <ComponentKit/CKComponentActionInternal.h>
 
 @class CKComponent;
 
-/**
- A component action is simply a selector. The selector may optionally take one argument: the sending component.
-
- Component actions provide a way for components to communicate to supercomponents using CKComponentActionSend. Since
- components are in the responder chain, the message will reach its supercomponents.
- */
-typedef SEL CKComponentAction;
-
 typedef NS_ENUM(NSUInteger, CKComponentActionSendBehavior) {
   /** Starts searching at the sender's next responder. Usually this is what you want to prevent infinite loops. */
   CKComponentActionSendBehaviorStartAtSenderNextResponder,
   /** If the sender itself responds to the action, invoke the action on the sender. */
   CKComponentActionSendBehaviorStartAtSender,
 };
 
+template<typename... T>
+struct CKTypedComponentAction {
+  CKTypedComponentAction() : _variant(CKTypedComponentActionVariantRawSelector), _target(nil), _selector(NULL) {};
+  CKTypedComponentAction(id target, SEL selector) : _variant(CKTypedComponentActionVariantTargetSelector), _target(target), _selector(selector)
+  {
+#if DEBUG
+    std::vector<const char *> typeEncodings;
+    _CKTypedComponentActionTypeVectorBuild(typeEncodings, _CKTypedComponentActionTypelist<T...>{});
+    _CKTypedComponentDebugCheckTargetSelector(target, selector, typeEncodings);
+#endif
```

Comment:
This little function does as much type-checking as we can do at the component action constructor site, which should be easier to understand when it fails on you than later, when actions are sent. I've seen this a few times where actions get out of sync, and we don't find out until the action fires. People sometimes don't test them.